### ISSUE TITLE
SerializationExceptions when updating an attribute to an empty custom string

### ISSUE THREAD

**Issue by finnbear** [guregu/dynamo#151]
I have a server that makes frequent use of your excellent `dynamo` package to save various data. When an error is encountered, I print it out with a statement such as `log.Println(err)`. Recently, I have been getting the error
```console
SerializationException: status code: 400, request id: O6MMFM7MPC2NI4DV407EI50J0JVV4KQNSO5AEMVJF66Q9ASUAAJG
```
The last time I got a `SerializationError`, it was due to #137 which was a bug with this package.

This time, the code throwing the error was performing an `Update` operation. I suspect that the problem was `update.Set`ing a string to an empty value. I will continue to look into the cause of the issue, but I am of the opinion that your library should catch this issue before it hits dynamo and either throw an preemptive and descriptive error, panic, or change the Update to a Delete (in my order of preference).

```go
    update := db.userTable.Update(userIndexID, userID)
    update.Set("color", user.Color) // TextMarshaler
    update.Set("experience", user.Experience) // int
    update.Set("money", user.Money) // int
    if len(user.Achievements) > 0 {
        update.SetSet("achievements", user.Achievements) // map[TextMarshaler]Empty (Empty is struct{})
    }
    update.Set("statistics", user.Statistics) // map[TextMarshaler]float64
    if len(user.CanonicalName) > 0 { // added this check today, I'm now seeing if the error still appears
        update.Set("canonicalName", string(user.CanonicalName)) // UserName (string)
    }
    if len(user.Name) > 0 { // added this check today, I'm now seeing if the error still appears
        update.Set("name", user.Name) // UserName (string)
    }
    if user.Updated > 0 {
        update.Set("updated", user.Updated) // int64
    }
    return update.Run()
```

Further research reveals that empty strings are now supported by dynamodb for non-key attributes (see https://aws.amazon.com/about-aws/whats-new/2020/05/amazon-dynamodb-now-supports-empty-values-for-non-key-string-and-binary-attributes-in-dynamodb-tables/). I am continuing to investigate the true cause of the error (it might be that the version of the AWS SDK used is not high enough to support empty strings).

Do you know of a better way to diagnose `SerializationException`s than blindly guessing what caused it? The problem I just described has gone unnoticed in my server's logs for the last 3+ days, possibly causing data loss for a subset of users, and I would like to avoid this type of problem in the future.

**Comment by guregu** [comment]
Hi, in the past I've experienced a similar error when I sent malformed UTF-8 strings to DynamoDB. Go doesn't check if a string is valid UTF-8, but DynamoDB will return a SerializationError when that happens. You can use [`utf8.ValidString`](https://godoc.org/unicode/utf8#ValidString) to check that. Also, if you use []byte instead of string it will happily accept invalid UTF-8.
By default this library won't set strings to empty on an Update unless you use SetNullable so I don't think empty strings are the problem.

In the invalid UTF8 case, the error message has the position that causes the error:
SerializationException: malformed input around byte 84 status code: 400, request id: 97C9AST0MJR2GV8TC13GJFJ917VV4KQNSO5AEMVJF66Q9ASUAAJG

If you use `LogLevel: aws.LogLevel(aws.LogDebugWithHTTPBody)` in your aws.Config you can dump all of your requests but it's a lot of logging.

Does your SerializationException just say "SerializationException" and not "malformed input", etc?

**Comment by guregu** [comment]
I ran a few tests on your other types (custom string types, TextMarshaler maps) but they came out fine 🤔

I think the UTF8 issue is a likely cause, if it’s possible to run a script that loops through all your users check all of your usernames for valid UTF8 that could give us the answer

**Comment by finnbear** [comment]
My error message was exactly as copied in my message, it contained no hints (partly why I described it as undiagnosable). I will try inserting utf8.ValidStirng checks on all strings and text marshalers (thanks for the hint!). I will also check that the keys of the map and set did not marshal to empty strings. What's weird is, according to the logs, the error stopped happening around the time I added the `if len(attribute) > 0` checks...

**Comment by finnbear** [comment]
@guregu Found it! To reproduce the problem, use an update to set an **empty** custom string:
```
    update := userTable.Update("userID", "testing")
    type customString string
    update.Set("customString", customString(""))
    err := update.Run()
    if err != nil {
        fmt.Println(err)
    }
}
```
Here is the line that I believe to be the problem:
https://github.com/guregu/dynamo/blob/20f2fb8805f070267b4ed8bf111f604828a1367a/substitute.go#L84
Right now, it can't handle custom strings properly.

**Comment by guregu** [comment]
Sweet, thanks. The types in ItemNamePlaceholder refer to the `$` used for expression names, Update.Set uses `?` (ItemValuePlaceholder) for values so that shouldn't be the issue. I will do some digging.

**Comment by guregu** [comment]
OK I've figured it out I think. Will get a fix in soon. Your findings were a lot of help.
The strange part is that the bug should result is an invalid value like `"ExpressionAttributeValues":{":v0": null}` but the AWS package encodes it like: `"ExpressionAttributeValues":{":v0":}` with the hanging colon, triggering the serialization error.
Fix should be up in a few minutes after I finish the tests.

**Comment by finnbear** [comment]
Excellent! As far as I'm concerned, there is no rush (minutes/hours until fix don't matter much to me). Putting in the len(attribute) > 0 checks has patched the immediate concern. I'm glad to here a better solution is on the way, though!

**Comment by guregu** [comment]
Just released v1.10.2 which fixes this.
The root of the problem was from 305af82c08899f4a28e89b7ac736ce933afcabb4 which attempted to fix this but was missing checks for typed strings (at the time DynamoDB couldn't handle empty strings at all, so attempting to set them would always cause an error).

I changed the logic so it attempts to encode the value first and checks that result instead of examining the type/value before encoding. Also added some additional checks so adding nils to ExpressionAttributeValues should always result in an error before it gets sent, to make it more clear so we don't get these confusing SerializationExceptions from AWS.

Thanks for the help and sorry about the issue.

**Comment by finnbear** [comment]
Sounds great! Thanks for developing and releasing a fix right away :1st_place_medal: 

I will pull the latest version and compile it into my development build, to address any issues before compiling it into production. **We can close this issue**, because I don't foresee any further problems.
